### PR TITLE
Allow ENVOY_LOG to log messages in user-specified logging level and format

### PR DESCRIPTION
**Pull request by jiajunye**
This is requested by https://github.com/envoyproxy/nighthawk/issues/290.

I added another constructor in ServiceImpl to accept an Envoy Logger Context pointer which contains user-specified log format and log level.

### CODE REVIEW

**Review by oschaaf — APPROVED** [review]
Nice; it is a shame we don't have an easy means to unit test this :-(
LGTM

**Comment by oschaaf** [comment]
The current change LGTM and makes sense, but I'm not sure I can fully match it to the PR description?

File: source/client/service_impl.h
Comment (oschaaf) [inline source/client/service_impl.h:38]: Wait .. one ask, does this then scan/parse the CLI for arguments that match Envoy's CLI in nighthawk_service? If so, wouldn't it be better to replicate these options in Nighthawk's own?

**Review by oschaaf — CHANGES_REQUESTED** [review]
Sorry -- one question, I think I need a little more context to understand this change.

**Comment by jiajunye** [comment]
> The current change LGTM and makes sense, but I'm not sure I can fully match it to the PR description?

Sorry for the confusion Otto. My previous PR title "Allow ENVOY_LOG to log messages in user specified log delegate" actually contains two parts: 1) in users' own logging framework, they need add some code to use their own log delegate. 2) in order to correctly display the messages in users' own log delegate, in the open source, we should enable the ability for users to choose the log level and format. This PR is just responsible for the part 2).

I've updated the title and description. Please let me know if this look better to you.

**Comment by oschaaf** [comment]
@jiajunye thanks for updating and sharing context, that is helpful, and now it makes sense to me.

File: source/client/service_impl.h
Comment (oschaaf) [inline source/client/service_impl.h:20]: Is this PR updated to the latest on master? I’m surprised the formatting check doesn’t object to this?
Regardless this change doesn’t seem nessecary?

File: source/client/service_impl.h
Comment (oschaaf) [inline source/client/service_impl.h:42]: Optional: using a && arg here might save a copy?

**Review by oschaaf — COMMENTED** [review]
A nit, a suggestion, and a question. Of it's not merges to recent master, could you do that? Otherwise LGTM

File: source/client/service_impl.h
Comment (jiajunye) [reply source/client/service_impl.h:42]: What's the benefit to choose && comparing to &?

File: source/client/service_impl.h
Comment (jiajunye) [reply source/client/service_impl.h:20]: Agreed. I will leave it unchanged.

Comment (jiajunye) [reply source/client/service_impl.h]: PR is updated to the latest on master.

File: source/client/service_impl.h
Comment (oschaaf) [reply source/client/service_impl.h]: Well, my take on it would be:
- With `&param` the smart pointer would be passed by reference, but ownership may or may not be taken by the constructor of `ServiceImpl`. The caller should make no assumptions about that and live with both scenarios after the call. 
- With a `const & std:unique_ptr<...>` parameter it would be clear that no ownership will be taken, but it might be better to unwrap and just pass a reference to the contained object itself as that avoids having to check for nullptr -- or at least delegates that responsibility to the caller.
- With `&&param` it is clear to consuming code that it will no longer own whatever it passes in to such a parameter (irrespective of wether there is a smart pointer involved). And `&&` saves a copy, which isn't a deciding factor in the code line we're talking about here (as it's not frequently called) -- but minding that might be a good habit.

Having said that, it may be good to add a test to cover this constructor, which will then probably also give our friend clang-tidy an opportunity to have a say about it when it inspects how this is going to be used :-) Next to adding actual coverage, which won't hurt either.

Comment (jiajunye) [reply source/client/service_impl.h]: Thanks so much Otto! This helps a lot!

File: test/service_test.cc
Comment (oschaaf) [inline test/service_test.cc:119]: Might want to make this a member instead of local variable to ensure the proper lifetime,
as the Logging::Context constructor takes a ref?:

```
  Context(spdlog::level::level_enum log_level, const std::string& log_format,
          Thread::BasicLockable& lock, bool should_escape, bool enable_fine_grain_logging = false);
```

Comment (jiajunye) [reply test/service_test.cc]: Agreed. I'm updating this PR to retest, since my local test can succeed somehow. Thanks Otto!

**Comment by jiajunye** [comment]
@oschaaf Is this integration test flaky? I had couple failed builds but one successful build.

**Comment by oschaaf** [comment]
> @oschaaf Is this integration test flaky? I had couple failed builds but one successful build.
I'm not sure if this is a flake, it is trending just below the threshold. Did you check the coverage report? Maybe there's some code that can be easily covered in a unit test, if not we need to lower the threshold. (the ratio may also change because of code churn).

**Comment by jiajunye** [comment]
> > @oschaaf Is this integration test flaky? I had couple failed builds but one successful build.
> 
> I'm not sure if this is a flake, it is trending just below the threshold. Did you check the coverage report? Maybe there's some code that can be easily covered in a unit test, if not we need to lower the threshold. (the ratio may also change because of code churn).

**Review by oschaaf — DISMISSED** [review]

**Comment by oschaaf** [comment]
(looks great from my perspective; thanks for addressing the feedback)

**Comment by jiajunye** [comment]
> (looks great from my perspective; thanks for addressing the feedback)

Thanks for reviewing the code. I really appreciate your help! Let's wait for dubious90's response on the threshold.

File: test/service_test.cc
Comment (dubious90) [inline test/service_test.cc:139]: Recognize that this is copy-paste of initial code, but I'm going to review the new code to match better style guide conventions.

First: this comment should be superfluous with a good test name.. We can remove that, and then change the test name.

Test names should ideally read like a sentence explaining what the test case is.  I'll defer to you on what makes sense for a name but typically it's something like:
ConstructorWithLoggingParameterCanRespondToRequests

File: test/service_test.cc
Comment (dubious90) [inline test/service_test.cc:142]: Two things here:
1. We try to avoid using auto for readability reasons, unless it is incredibly obvious what the returning type is (such as when the constructor already indicates the type or we're iterating through an array where the type is immediately present there).  See the "Decision" at https://google.github.io/styleguide/cppguide.html#Type_deduction

2. We also try to avoid abbreviations in general.  If this is a request, then we'd usually call the variable request instead of r.  See https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: test/service_test.cc
Comment (dubious90) [inline test/service_test.cc:150]: auto critique applies here too, especially since there are two different statuses - rpc status and absl status, so it's important to distinguish which one is here.

**Review by dubious90 — CHANGES_REQUESTED** [review]
Hi Jason, mostly looks good with a couple style comments.  Please reach out if you have any questions/concerns.

File: test/service_test.cc
Comment (jiajunye) [reply test/service_test.cc:133]: I agree that this should be private. The reason it needs to be a member variable is that the logging_context constructor takes log_lock_ by ref.

**Comment by dubious90** [comment]
> You're right. We split the integration tests and unit tests, and while the unit test covers the new constructor, the integration tests do not. I'm think that lowering the threshold is what we should do in this case, if @dubious90 agrees. That threshold can be tweaked in `ci/do_ci.sh`.

I agree with Otto here.  I think I might have further questions about how the coverage works now, but that is separate from this PR.

File: test/service_test.cc
Comment (jiajunye) [reply test/service_test.cc:139]: It looks like we typically match test suite name to the class name. ConstructorWithLoggingParameterCanRespondToRequests sounds like a good test name in TEST_P:
http://google3/third_party/googletest/googletest/include/gtest/gtest-param-test.h?l=416&rcl=317696457

Comment (jiajunye) [reply test/service_test.cc:139]: Oh, I guess you mean the line 141. Now I get it.

File: test/service_test.cc
Comment (dubious90) [inline test/service_test.cc:145]: Ah, I actually gave you a bad name for this. This isn't actually a request.  Sorry about that!  Easier to tell what it is now that it's not an `auto`.

Whatever makes sense to you as a name here.  reader_writer is an option.  Or if you don't like that, something that describes this accurately, like stream_client or stream_handler.  Whatever you feel good about.

File: test/service_test.cc
Comment (dubious90) [inline test/service_test.cc:143]: Note that we are actually within namespace ::nighthawk::client, so you can just say ExecutionRequest and ExecutionResponse with no namespace declarations.  Should make this more concise.

**Review by dubious90 — COMMENTED** [review]
Probably a last round of requests.  Thanks for your work!

File: test/service_test.cc
Comment (jiajunye) [reply test/service_test.cc:145]: I would go with stream. Thanks for the advice!

File: test/service_test.cc
Comment (jiajunye) [reply test/service_test.cc:143]: It actually complains about _Unknown type name 'ExecutionRequest'; did you mean 'nighthawk::client::ExecutionRequest'?_ after removing ::nighthawk::client. Maybe because I added an empty namespace wrapper?

Comment (jiajunye) [reply test/service_test.cc]: Talked in person with @dubious90. The error was caused by the namespace discrepancy between the code file and proto buffer. In code, our namespaces start with uppercase, while in proto buffer they start with lowercase.

**Review by dubious90 — APPROVED** [review]
LGTM.  Thanks for responding to all these changes!

**Comment by jiajunye** [comment]
@dubious90 Do you mind merging this PR please? Thanks!

**Pull request merged**
